ETH-Merge-Bots/Changes-in-Block-Difficulty/src/agent.ts

```diff
@@ -0,0 +1,74 @@
+import { Finding, HandleBlock, BlockEvent, getEthersProvider, ethers } from "forta-agent";
+import { NUMBER_OF_BLOCKS_TO_CHECK, FINDING_THRESHOLD, createFinding, TERMINAL_TOTAL_DIFFICULTY } from "./utils";
+import BigNumber from "bignumber.js";
+
+BigNumber.set({ DECIMAL_PLACES: 2 });
+
+export const provideHandleBlock = (
+  init: boolean,
+  provider: ethers.providers.Provider,
+  numberOfBlocksToCheck: number,
+  threshold: BigNumber
+): HandleBlock => {
+  let blockDifficulties: BigNumber[] = new Array<BigNumber>();
+  let blockDifficultyMovingAverage: BigNumber;
+  let blockDifficultiesSum: BigNumber = new BigNumber(0);
+
+  return async (blockEvent: BlockEvent) => {
+    const findings: Finding[] = [];
+
+    // check if TTD has been reached
+    if (blockEvent.block.totalDifficulty >= TERMINAL_TOTAL_DIFFICULTY) return findings;
+
+    // executed on the first run
+    if (!init) {
+      init = true;
+      //populate the blockDifficulties array in ascending block order
+      for (let i = 0; i <= numberOfBlocksToCheck - 1; i++) {
+        const blockDifficulty = new BigNumber(
+          (await provider.getBlock(blockEvent.blockNumber - numberOfBlocksToCheck + i))._difficulty.toString()
+        );
```

Comment:
since there are 100 blocks that uses provider, would it make sense if we add `try/catch` here and if there's no error, making `init = true`?

Reply:
I think `try/catch` should only be used in calls that can actually fail, i.e. calls to non-existing addresses or non-existing contract functions. As there are multiple nodes, I think we can, in principle, take for granted that this is going to work/the nodes use the "best" rpcUrl available, or at least that's what I've got in mind:D

Reply:
okay because it takes around 2 minutes that's why I was worried but we can monitor the deployment :D 👍

Reply:
Hmm, this is too much, can it be a connection issue? It's closer to 25-30" for me using the default endpoint

Reply:
oh definitely a country issue then 😄 😭